src/transformers/models/kosmos2/modeling_kosmos2.py

```diff
@@ -1149,7 +1149,9 @@ def forward_embedding(
             inputs_embeds = self.embed_tokens(input_ids)
 
         if image_embeds is not None:
-            inputs_embeds[img_input_mask.to(dtype=torch.bool)] = image_embeds.view(-1, image_embeds.size(-1))
+            inputs_embeds[img_input_mask.to(dtype=torch.bool)] = image_embeds.to(inputs_embeds.device).view(
+                -1, image_embeds.size(-1)
+            )
```

Comment:
@muellerzr As `forward_embedding` is a direct child of a big component `Kosmos2TextTransformer`, I can't put `Kosmos2TextTransformer` in `_no_split_modules` IIRC.

So the best we can do is just to put the tensors in the necessary device manually like done here, right?

In general, it seems to me  we have to do this at every operation (which is tedious and so ugly), but I am not sure. Do you have any insight to share here?

Reply:
I'm not sure I get `image_embeds.to(inputs_embeds.device)` - is `image_embeds` distributed across different devices? 

For `_no_split_modules` - what is the size of `Kosmos2TextTransformer`? In retrospect, we might have wanted a `Kosmos2Embedding` class to enable this. One possibility would be enabling specifiying specific layers e.g. `_no_split_modules = ["Kosmos2TextTransformer.embed_tokens"]`

Reply:
> is image_embeds distributed across different devices
Here `image_embeds` is computed by the vision component, and vision component may have weights on other devices than the text component.

> Kosmos2Embedding class

`Kosmos2Embedding` --> This would break the pretrained weight loading, but yes, in general I agree. We have some models with specific embedding layers , but others not.

Reply:
> Here image_embeds is computed by the vision component, and vision component may have weights on other devices than the text component. @muellerzr will have more expertise than I can share :-)

I still don't understand 😅  Why does `x.to(x.device)` work in this case? If `input_embeds` is computed by the vision component, would this keep it on the same device as the vision component? 

> In general, it seems to me we have to do this at every operation (which is tedious and so ugly),

Regarding your initial comment, if I'm operating in non-distrubted model is this still an expensive operation?

Reply:
Hi, let me double check we are in the same track: here is `image_embeds.to(inputs_embeds.device)`: `inputs_embed` is the `text` (embedding of `input_ids` computed by text component, and `image_embeds` is computed by the vision component.

> is this still an expensive operation?
If they are on the same already device, the `to(device)` is a no-op I believe (i.e. cost nothing)

Reply:
d'oh 🤦‍♀️ - sorry - I was completely misreading this. In that case - all good!